### DISCUSSION TITLE
text outline option for DrawText()

### DISCUSSION THREAD

**Discussion by imm0r** [Spawnova/ShinsOverlayClass#4] [Ideas]
First of all I want to congratulate you on a really great library! Awesome job you did there.
But 'again' I have to state that your gdip texts don't have an option for colored outlines ;)
The drop shadow option is a nice substitute but not really a proper outline.

I know it's already done in several libraries but it would be nice to see your DrawText function offering the same functionality.

`DllCall("gdiplus\GdipAddPathString", Ptr, pPath, "WStr", String, "int", -1, Ptr, hFontFamily, "int", Style, "float", Size, Ptr, &RectF, Ptr, hStringFormat)`

_Looks like the most promising approach... was also used [here](https://github.com/Ixiko/AHK-libs-and-classes-collection/blob/master/libs/g-n/Gdip_All.ahk)._

**Comment by Spawnova** [comment]
Hey thanks, unfortunately I don't see a convenient way to add true outlines to the text currently, at least the way I am doing it for the sake of performance, eventually I'm just going to release a dll that has more support in general. I've added an outline option for now anyways but it's just drawing the text 8 times to create a border.

A real solution requires using glyphs as far as I know but I'm going to utilize that in a dll instead at some point.

**Reply by imm0r** [reply]
[This one](https://github.com/imm0r/AHK-Gdip_TextToGraphics2/blob/de32297d98dacc8c58632f3afbfb79f1154ca197/libs/g-n/Gdip_Ext.ahk) is using **`gdiplus\GdipAddPathString`** for outlining a string...
I'm not entirely sure if this approach uses _glyphs_, as I don't quite understand what exactly you mean by _glyphs_.

But yes, you are absolutely right. Drawing the text 8 times to create a border is in no way convenient or even performant. 😎

**Reply by Spawnova** [reply]
This library is focused around direct2d, unfortunately it doesn't support gdi/gdi+ in most cases, it's also quite limited since it's written purely in ahk. I can do a lot more in c++ but it requires the user to use a supplied dll as well, so in the future I will support this lightweight version, and a more extensive dll version. =P

d2d is already very fast so I'm sure the simple outline function performance shouldn't even be noticeable until you start using several hundred DrawText() calls per frame, which doesn't seem like something anyone would ever reach lol

But either way, real solutions will be utilized in a compiled dll instead, built specifically to be used with ahk.
